File: backend-node/src/lib/plans/planCatalogPromptCap.js
/**
 * Trim merged food/exercise catalogs before sending to FastAPI / Claude.
 * Full catalogs are still used for bind/validation after generation.
 */

function readIntEnv(name, fallback) {
  const n = Number(process.env[name]);
  return Number.isFinite(n) && n > 0 ? Math.floor(n) : fallback;
}

function trimFoodsForPrompt(foods, opts = {}) {
  const list = Array.isArray(foods) ? foods : [];
  const maxTotal = opts.maxTotal ?? readIntEnv('PLAN_PROMPT_MAX_FOODS', 80);
  const maxPerGroup = opts.maxPerGroup ?? readIntEnv('PLAN_PROMPT_MAX_FOODS_PER_GROUP', 8);
  const groupCounts = {};
  const out = [];

  for (const item of list) {
    const group = item.planGroup || 'other';
    groupCounts[group] = groupCounts[group] || 0;
    if (groupCounts[group] >= maxPerGroup) continue;
    groupCounts[group] += 1;
    out.push(item);
    if (out.length >= maxTotal) break;
  }
  return out;
}

function exerciseCellKey(item) {
  const muscle = item.muscleGroup || 'other';
  const difficulty = item.planDifficulty || item.difficulty || 'intermediate';
  return `${muscle}:${difficulty}`;
}

function trimExercisesForPrompt(exercises, opts = {}) {
  const list = Array.isArray(exercises) ? exercises : [];
  const maxTotal = opts.maxTotal ?? readIntEnv('PLAN_PROMPT_MAX_EXERCISES', 120);
  const maxPerCell = opts.maxPerCell ?? readIntEnv('PLAN_PROMPT_MAX_EXERCISES_PER_CELL', 6);
  const cellCounts = {};
  const out = [];

  for (const item of list) {
    const key = exerciseCellKey(item);
    cellCounts[key] = cellCounts[key] || 0;
    if (cellCounts[key] >= maxPerCell) continue;
    cellCounts[key] += 1;
    out.push(item);
    if (out.length >= maxTotal) break;
  }
  return out;
}

function trimPlanCatalogForPrompt({ foods, exercises } = {}) {
  return {
    foods: trimFoodsForPrompt(foods),
    exercises: trimExercisesForPrompt(exercises),
  };
}

module.exports = {
  trimFoodsForPrompt,
  trimExercisesForPrompt,
  trimPlanCatalogForPrompt,
  exerciseCellKey,
};

File: ai-service/app/services/plan_catalog_prompt_cap.py
"""
Cap food/exercise lists before building the Claude plan prompt.
"""

from __future__ import annotations

from typing import Any

from app.config import get_settings


def _exercise_cell_key(item: dict[str, Any]) -> str:
    muscle = str(item.get("muscleGroup") or "other")
    difficulty = str(item.get("planDifficulty") or item.get("difficulty") or "intermediate")
    return f"{muscle}:{difficulty}"


def trim_foods_for_prompt(foods: list[dict[str, Any]]) -> list[dict[str, Any]]:
    settings = get_settings()
    max_total = max(20, int(settings.plan_prompt_max_foods or 80))
    max_per_group = max(3, int(settings.plan_prompt_max_foods_per_group or 8))
    group_counts: dict[str, int] = {}
    out: list[dict[str, Any]] = []
    for item in foods:
        group = str(item.get("planGroup") or "other")
        count = group_counts.get(group, 0)
        if count >= max_per_group:
            continue
        group_counts[group] = count + 1
        out.append(item)
        if len(out) >= max_total:
            break
    return out


def trim_exercises_for_prompt(exercises: list[dict[str, Any]]) -> list[dict[str, Any]]:
    settings = get_settings()
    max_total = max(30, int(settings.plan_prompt_max_exercises or 120))
    max_per_cell = max(3, int(settings.plan_prompt_max_exercises_per_cell or 6))
    cell_counts: dict[str, int] = {}
    out: list[dict[str, Any]] = []
    for item in exercises:
        key = _exercise_cell_key(item)
        count = cell_counts.get(key, 0)
        if count >= max_per_cell:
            continue
        cell_counts[key] = count + 1
        out.append(item)
        if len(out) >= max_total:
            break
    return out


def trim_plan_catalog_for_prompt(
    foods: list[dict[str, Any]],
    exercises: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    return trim_foods_for_prompt(foods), trim_exercises_for_prompt(exercises)
